### PR TITLE
FilterOp support + Using a different variant type for the evaluation of ExpressionTree

### PR DESCRIPTION
**Pull request by bowu**
The FilterOp's implementation is very basic. It computes a var binding map for each row of the input table and evaluate the ExpressionTree associated with the predicate to determine whether the row should be pruned.

ExpressionTreeLiteral is the variant type used to build the expression tree, but it has QueryProperty as an alternative type, which makes it difficult to use in the operators. VariantType, which only has plain C++ types, is now used for ExpressionTree evaluation.

### CODE REVIEW

**Comment by l-hoang** [comment]
meta level comment: you can avoid the long list of older commits by checking out the latest changes from master and building your branch on top of that since master will have all merged PRs up to a certain point

File: libquery/include/galois/graphs/AttributedGraph.h [hunk 1]
```diff
@@ -71,12 +60,18 @@ class GALOIS_EXPORT AttributedGraph {
...
-  std::optional<AttributeType> GetEdgeAttribute(uint64_t eid,
-                                                std::string attribute_name);
+  std::optional<VariantType> GetEdgeAttribute(uint64_t eid,
+                                              std::string attribute_name);
+
+  //! Return the arrow array for some node attribute
+  const IndexableChunkedArray& GetNodeAttributeArray(std::string attribute);
+
+  //! Return the arrow array for some edge attribute
+  const IndexableChunkedArray& GetEdgeAttributeArray(std::string attribute);
```
Comment (danielmawhirter) [inline libquery/include/galois/graphs/AttributedGraph.h:74]: See https://github.com/KatanaGraph/katana/pull/432#discussion_r478571924

File: libquery/include/querying/Table.h [hunk 2]
```diff
@@ -89,12 +94,13 @@ class Table : public InterData {
   GetColumnByName(std::string name);
 
   Schema GetSchema() const;
+  void SetSchema(galois::Table::Schema schema);
```
Comment (danielmawhirter) [inline libquery/include/querying/Table.h:97]: Where is this needed?

File: libquery/src/ExpressionTree.cpp [hunk 3]
```diff
@@ -143,9 +162,11 @@ ExpressionTreeLiteral ArithmeticTree::Evaluate(const VarBindingMap& vars) {
   } else if (variant_index == 0) {
     // case through the different arithmetic function types and evaluate
     // as necessary; combine left/right according to operator
-    ExpressionTreeLiteral left_result  = left_->Evaluate(vars);
-    ExpressionTreeLiteral right_result = right_->Evaluate(vars);
+    VariantType left_result  = left_->Evaluate(vars);
+    VariantType right_result = right_->Evaluate(vars);
 
+    std::cout << "indices: " << left_result.index() << ", "
+              << right_result.index();
```
Comment (danielmawhirter) [inline libquery/src/ExpressionTree.cpp:169]: GALOIS_LOG_VERBOSE?

Comment (bowu) [reply libquery/src/ExpressionTree.cpp:169]: I'll delete this.

File: libquery/include/galois/graphs/AttributedGraph.h (hunk 1 above)
Comment (bowu) [reply libquery/include/galois/graphs/AttributedGraph.h:74]: What should we do about this then? @l-hoang

File: libquery/src/OpGraphBuilder.cpp [hunk 4]
```diff
@@ -235,11 +236,17 @@ void OpGraphBuilder::AddOp(std::unique_ptr<Operator> new_op) {
...
+                                CypherCompiler& ast_parser) {
+  std::unique_ptr<MatchOp> new_match_op = std::make_unique<MatchOp>(
       bit_label_schema, ast_parser.GetQueryNodes(), ast_parser.GetQueryEdges());
   GALOIS_LOG_DEBUG("Adding MATCH op");
-  AddOp(std::move(new_op));
+  AddOp(std::move(new_match_op));
+  if (!ast_parser.GetExpressionTrees().empty()) {
+    std::unique_ptr<FilterOp> new_filter_op =
+        std::make_unique<FilterOp>(ast_parser.GetExpressionTrees());
+    ast_parser.ResetExpressionTrees();
```
Comment (danielmawhirter) [inline libquery/src/OpGraphBuilder.cpp:247]: If these are modifiable, they should be moved out of CypherCompiler IMO

File: libquery/src/Table.cpp [hunk 5]
```diff
@@ -277,10 +276,14 @@ galois::Table::GetColumnByName(std::string name) {
       return std::make_pair(schema_.GetType(col), GetColumn(col));
     }
   }
+  std::cout << table_->ToString();
```
Comment (danielmawhirter) [inline libquery/src/Table.cpp:279]: I don't believe this should still be here

**Comment by bowu** [comment]
> meta level comment: you can avoid the long list of older commits by checking out the latest changes from master and building your branch on top of that since master will have all merged PRs up to a certain point

Can I still do it for this branch? I guess the PR is bound to the branch.

**Comment by roshandathathri** [comment]
> Can I still do it for this branch? I guess the PR is bound to the branch.

Yes, you can. I don't know your setup but assume "master" is katana master and "dev" is your branch (tied to this PR) on your fork "fork" of master. Then you can do this:
`git checkout master
git pull
git checkout dev
git rebase -i master
git push --force fork dev`
If it's not clear and if there are issues, you can message me directly.

File: libquery/src/ExpressionTree.cpp [hunk 6]
```diff
@@ -5,86 +5,95 @@
...
-      break;
-    case kArithModulo:
-      return_result.emplace<2>(std::get<2>(left_result) %
-                               std::get<2>(right_result));
-      break;
-    default:
-      GALOIS_LOG_FATAL("Unsupported arithemitc operation on integers");
-      break;
-    }
+  if (variant_index == 0) {
```
Comment (l-hoang) [inline libquery/src/ExpressionTree.cpp:77]: case statement can be used here

File: libquery/src/ExpressionTree.cpp [hunk 7]
```diff
@@ -5,86 +5,95 @@
...
 
 ////////////////////////////////////////////////////////////////////////////////
 // Arithmetic Tree
 ////////////////////////////////////////////////////////////////////////////////
 
-ExpressionTreeLiteral
-ArithmeticTree::EvaluateLiterals(ExpressionTreeLiteral left_result,
-                                 ExpressionTreeLiteral right_result) {
+template <int IND>
+VariantType ArithmeticTree::IntegerFunc(ArithmeticTreeOperator op,
```
Comment (l-hoang) [inline libquery/src/ExpressionTree.cpp:17]: Very neat way of doing this

File: libquery/include/querying/ExpressionTree.h [hunk 8]
```diff
@@ -113,14 +116,21 @@ class ArithmeticTree {
   // TODO function constructor: takes a string (function name) and a list of
   // arguments
 
+  template <int IND>
+  VariantType IntegerFunc(ArithmeticTreeOperator op, VariantType left,
+                          VariantType right);
+
+  template <int IND>
+  VariantType FloatFunc(ArithmeticTreeOperator op, VariantType left,
```
Comment (roshandathathri) [inline libquery/include/querying/ExpressionTree.h:124]: Did this build? I'm surprised by that. Template function definitions are required to be in the header if they are used anywhere outside the file that they are defined in.

Template parameter name and type are confusing here. Is it the OpIndex or the VariantIndex? In both cases, uint32_t or size_t is better for the type of the index.

After looking at the definitions, I think these two functions can be merged into a single function with OperationIndex and VariantIndex as template or regular parameters. I'm not sure how it used though to comment on whether template parameters are useful.

File: libquery/include/querying/OpGraphBuilder.h [hunk 9]
```diff
@@ -30,7 +30,7 @@ class GALOIS_EXPORT OpGraphBuilder {
   //! This will involve at least a MATCH node and possibly a FILTER node
   //! to operation graph.
   void AddMatchOp(const galois::BitLabelSchema& bit_label_schema,
-                  const CypherCompiler& ast_parser);
+                  CypherCompiler& ast_parser);
```
Comment (roshandathathri) [inline libquery/include/querying/OpGraphBuilder.h:33]: Why was this change required? Ideally, it just be a read, so "const" is appropriate.

File: libquery/include/querying/Table.h [hunk 10]
```diff
@@ -8,6 +8,8 @@
 
 namespace galois {
 
+class Table;
```
Comment (roshandathathri) [inline libquery/include/querying/Table.h:11]: Why was this required?

File: libquery/include/querying/Table.h (hunk 2 above)
Comment (roshandathathri) [reply libquery/include/querying/Table.h:97]: Also "Schema" is sufficient and "galois::Table::" is superfluous.

File: libquery/src/FilterOp.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
+#include "querying/ops/FilterOp.h"
+#include "querying/Table.h"
+
+std::vector<galois::Operator*> galois::FilterOp::Compute() {
+  std::shared_ptr<Table> input = std::dynamic_pointer_cast<Table>(GetInput(0));
+  auto properties              = expression_trees_[0]->RequiredProperties();
```
Comment (l-hoang) [inline libquery/src/FilterOp.cpp:6]: assert that there is indeed only 1 expression tree at this point in time (will likely change later, but never hurts to be safe)

File: libquery/include/querying/ops/FilterOp.h [hunk 12]
```diff
@@ -6,11 +6,19 @@ namespace galois {
 
 class FilterOp final : public Operator {
 public:
-  std::vector<Operator*> Compute() override { return PropagateOutput(); }
+  FilterOp(std::vector<std::unique_ptr<ExpressionTree>>& expression_trees) {
+    std::move(expression_trees.begin(), expression_trees.end(),
+              std::back_inserter(expression_trees_));
```
Comment (roshandathathri) [inline libquery/include/querying/ops/FilterOp.h:11]: I think the below simpler code should work.
`FilterOp(std::vector<std::unique_ptr<ExpressionTree>>&& expressions_trees) : expression_trees_(std::move(expression_tree)) {}
`
Does it not work?

File: libquery/src/FilterOp.cpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        map.emplace(prop.first, table_data);
+        break;
+      }
+      case (Table::Type::STRING): {
+        auto arrow_array = std::static_pointer_cast<arrow::BinaryArray>(
+            type_array_pair.second);
+        std::string table_data = arrow_array->GetString(i);
+        map.emplace(prop.first, table_data);
+        break;
+      }
```
Comment (l-hoang) [inline libquery/src/FilterOp.cpp:67]: `map.emplace(prop.first + "$")`
the code still assumes that there is a $ if i recall correctly

File: libquery/include/galois/graphs/AttributedGraph.h (hunk 1 above)
Comment (l-hoang) [reply libquery/include/galois/graphs/AttributedGraph.h:74]: i would prefer not to expose the arrays at all unless you can pair it up with the reindexing array

the original idea was to resort the edge attribute array when the edges were sorted and abandon the edge reindexing array entirely. this will fix all the problems with this array currently, and it should be moved to that kind of flow later

for now, just fixing the get edge attribute function and using only that to access edge attrbutes should be fine from this PR's perspective, but from the perspective of direct access to the edge attribute array some other change will be required

File: libquery/include/querying/ExpressionTree.h [hunk 14]
```diff
@@ -113,14 +116,21 @@ class ArithmeticTree {
   // TODO function constructor: takes a string (function name) and a list of
   // arguments
 
+  template <int IND>
```
[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:119]: actually i'm not sure how that would play with the compiler since it's a template function, so maybe it needs to be in the header (not necessarily as a class function; I leave that up to you)

File: libquery/include/querying/ExpressionTree.h [hunk 15]
```diff
@@ -28,12 +28,15 @@ using VarPropPair = std::pair<std::string, std::string>;
 //! TODO may make more sense to extend int and float to 64bit versions
 using ExpressionTreeLiteral =
     std::variant<galois::QueryProperty, std::string, int, float, bool>;
+//! The possible types for runtime values
+using VariantType = std::variant<bool, std::string, uint32_t, int32_t, uint64_t,
```
[the remark this answers is not in the document]
Comment (danielmawhirter) [reply libquery/include/querying/ExpressionTree.h:32]: arrow::Scalar is an option

File: libquery/src/AttributedGraph.cpp [hunk 16]
```diff
@@ -227,7 +227,7 @@ AttributedGraph::GetNodeAttribute(uint32_t vid, std::string attribute_name) {
   return std::nullopt;
 }
 
-std::optional<AttributeType>
+std::optional<VariantType>
 AttributedGraph::GetEdgeAttribute(uint64_t eid, std::string attribute_name) {
```
[the remark this answers is not in the document]
Comment (danielmawhirter) [reply libquery/src/AttributedGraph.cpp:231]: Bo, it might be possible to use 2 levels of Take to resolve the other issue. For this case, my understanding is that the lookups are individual

File: libquery/src/OpGraphBuilder.cpp (hunk 4 above)
Comment (roshandathathri) [reply libquery/src/OpGraphBuilder.cpp:247]: Why does it need to be modified?

File: libquery/include/querying/ExpressionTree.h (hunk 15 above)
Comment (roshandathathri) [inline libquery/include/querying/ExpressionTree.h:32]: I find the name to not be meaningful. Is this the type of attributes or expression values?

File: libquery/include/querying/OpGraphBuilder.h (hunk 9 above)
Comment (l-hoang) [reply libquery/include/querying/OpGraphBuilder.h:33]: Bo is clearing the expression tree holder in the compiler object

now that you mention it, I believe the init call should already be doing that, so it may be unnecessary

File: libquery/include/querying/Table.h (hunk 10 above)
Comment (danielmawhirter) [reply libquery/include/querying/Table.h:11]: my oops, Builder and Schema used to be outside, and before, Table

**Comment by bowu** [comment]
> > Can I still do it for this branch? I guess the PR is bound to the branch.
> 
> Yes, you can. I don't know your setup but assume "master" is katana master and "dev" is your branch (tied to this PR) on your fork "fork" of master. Then you can do this:
> `git checkout master`
> `git pull`
> `git checkout dev`
> `git rebase -i master`
> `git push --force fork dev`
> If it's not clear and if there are issues, you can message me directly.

Thanks for the note. It causes multiple conflicts. I guess I'll create new branches for my future PRs.

File: libquery/include/galois/graphs/AttributedGraph.h (hunk 1 above)
Comment (roshandathathri) [reply libquery/include/galois/graphs/AttributedGraph.h:74]: Same comment as I left in the other thread:

In this case, the arrow array shouldn't be exposed outside the class as-is until we reindex the arrays. Either each access to the array should go through the AttributedGraph or we can create a new class that has both the array and the index-map, and the AttributeGraph can expose an object of this class.

File: libquery/include/querying/ExpressionTree.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/include/querying/ExpressionTree.h:119]: Yes, as its compiling currently, I think moving the declaration to the cpp file is better.

File: libquery/src/Table.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (roshandathathri) [reply libquery/src/Table.cpp:279]: In any case, we shouldn't be using cout and cerr directly. GALOIS_LOG_* is preferred.

[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/src/Table.cpp:279]: cout/cerr still have their uses if you want to print things that a user would want to read. the disadvantage of the galois_log functions is that they print a line number/file as well which to an average user would mean very little

File: libquery/include/galois/graphs/AttributedGraph.h (hunk 1 above)
Comment (l-hoang) [reply libquery/include/galois/graphs/AttributedGraph.h:74]: i'm leaving it public for now since otherwise it will break some code, but i am adding a comment saying they shouldn't be exposed as is

File: libquery/include/querying/ExpressionTree.h (hunk 8 above)
Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:124]: I think having separate float/int functions is fine as the ops you can do on them are different. Similarly having a different string function (if/when those get added) is fine IMO too.

File: libquery/include/querying/ops/FilterOp.h (hunk 12 above)
Comment (l-hoang) [reply libquery/include/querying/ops/FilterOp.h:11]: Actually, I just thought about it some more while making the change, and you probably don't want to do this because you want the vector object itself to remain owned by the CypherCompiler object so that it can clear it, so what Bo has here is probably best.

File: libquery/src/OpGraphBuilder.cpp (hunk 4 above)
Comment (l-hoang) [reply libquery/src/OpGraphBuilder.cpp:247]: I'm making it a pointer because the filter op constructor will be moving unique ptrs from the object (sounds like a modification to me!)

File: libquery/src/OpGraphBuilder.cpp [hunk 17]
```diff
@@ -235,11 +236,17 @@ void OpGraphBuilder::AddOp(std::unique_ptr<Operator> new_op) {
...
+    AddOp(std::move(new_filter_op));
+  }
```
[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/src/OpGraphBuilder.cpp:249]: decided to keep it in this function and opt for renaming addmatchop to matchop**s**; same with returnop -> returnops.

reasoning is they both can add more than one op node associated with match/return to the opgraph

File: libquery/include/querying/OpGraphBuilder.h (hunk 9 above)
Comment (l-hoang) [reply libquery/include/querying/OpGraphBuilder.h:33]: changed it to a pointer because unique_ptrs are being moved from the object by the filter op constructor

File: libquery/include/querying/ExpressionTree.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:32]: renamed ExpressionValueType and added enum indices

Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:32]: name changed

**Comment by l-hoang** [comment]
as Bo is busy I took the liberty of making the changes for him. @bowu take a look

the tests for the filter op will come in a later PR

**Comment by l-hoang** [comment]
pushed 2 commits fixing a bug in the expression tree exposed by a functioning filter op + fixed a bug with filter op not going over all expression trees under certain conditions

i'll rebase-interactive and reorder/squash commits around later

File: libquery/src/ExpressionTree.cpp [hunk 18]
```diff
@@ -5,85 +5,110 @@
...
+  case kArithMultiply:
+    return_result.emplace<type_index>(std::get<type_index>(left) *
+                                      std::get<type_index>(right));
+    break;
+  case kArithDivide:
+    return_result.emplace<type_index>(std::get<type_index>(left) /
+                                      std::get<type_index>(right));
+    break;
+  default:
+    GALOIS_LOG_FATAL("Unsupported arithmetic operation on floats");
```
Comment (roshandathathri) [inline libquery/src/ExpressionTree.cpp:73]: As type_index is a template parameter, I'm not sure why IntegerFunc and FloatFunc cannot be merged? If type index is float and the arithmetic operation is modulo, then the function can call fatal.

Comment (l-hoang) [reply libquery/src/ExpressionTree.cpp:73]: 🤔 

good point, will see if I can fix

Comment (roshandathathri) [reply libquery/src/ExpressionTree.cpp:73]: It's fine to not handle this in this PR.

File: libquery/include/querying/ExpressionTree.h [hunk 19]
```diff
@@ -23,17 +23,39 @@ namespace galois {
...
-//! Literals (except for index 0 which is a variable/property) that are
-//! returned during subtree evaluation
-//! TODO may make more sense to extend int and float to 64bit versions
-using ExpressionTreeLiteral =
+//! Type of an unevaluated expression stored in an expresson tree. Upon
+//! evaluation, it resolves into a more concrete type.
+using ExpressionAbstractType =
     std::variant<galois::QueryProperty, std::string, int, float, bool>;
+//! Enum to make indexing into ExpressionAbstractType more readable
+enum ExpressionAbstractIndex {
```
Comment (ddn0) [inline libquery/include/querying/ExpressionTree.h:31]: some preference for `enum class` over `enum`. This will give you more typed behavior and let the compiler warn you more strongly if you compare values from different enum classes: https://godbolt.org/z/ocjbhq

File: libquery/src/ExpressionTree.cpp (hunk 18 above)
Comment (l-hoang) [reply libquery/src/ExpressionTree.cpp:73]: i need to reorder some commits anyways, so i'll have to go through another round of CI anyways

File: libquery/include/querying/IndexableChunkedArray.h [hunk 20]
```diff
@@ -54,6 +54,10 @@ class IndexableChunkedArray {
   //! Returns number of elements in the chunked array
   int64_t size() const;
 
+  std::shared_ptr<arrow::ChunkedArray> GetOriginal() const {
```
Comment (ddn0) [inline libquery/include/querying/IndexableChunkedArray.h:57]: `const shared_ptr&`?

File: libquery/include/querying/ops/FilterOp.h [hunk 21]
```diff
@@ -6,11 +6,23 @@ namespace galois {
 
 class FilterOp final : public Operator {
 public:
-  std::vector<Operator*> Compute() override { return PropagateOutput(); }
+  FilterOp(std::vector<std::unique_ptr<ExpressionTree>>* expression_trees) {
+    // must first resize to same size as passed in tree vector
+    expression_trees_.resize(expression_trees->size());
+    // move the expression trees (note you don't move the vector itself)
+    std::move(expression_trees->begin(), expression_trees->end(),
```
Comment (ddn0) [inline libquery/include/querying/ops/FilterOp.h:13]: looks like there are some comments that I made in #438 that apply here as well.

This move doesn't do anything by itself.

File: libquery/src/ExpressionTree.cpp (hunk 18 above)
Comment (roshandathathri) [reply libquery/src/ExpressionTree.cpp:73]: Then it's fine :)

File: libquery/include/querying/ops/FilterOp.h (hunk 21 above)
Comment (l-hoang) [reply libquery/include/querying/ops/FilterOp.h:13]: oops, so i didn't fix it

`std::move(std::move(....), std::move(....), begin)` is the way to go then?

File: libquery/include/querying/IndexableChunkedArray.h (hunk 20 above)
Comment (l-hoang) [reply libquery/include/querying/IndexableChunkedArray.h:57]: got it

File: libquery/include/querying/ExpressionTree.h (hunk 19 above)
Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:31]: i didn't know that that existed/there was a difference; will do

**Comment by l-hoang** [comment]
> I have some questions about the general approach and its performance (versus arrow/compute/api.h) but that can be decided outside the context of this PR.

I'm assuming arrow compute can be used even when non-arrow arrays are involved?

and yes, performance wasn't really on our mind when we designed it this way. in particular, we did have some concerns about the way variables are being bound in the expression trees, but it looks like there's more than that

> Some unit tests of evaluation would be nice.

got you covered (......i think, unless you meant something else)

https://github.com/KatanaGraph/katana/blob/028d803661a52e7d1994d79b23967e0bb6ce49df/libquery/test/cypher-compile-test.cpp#L898

some tests to test the filter op's operation itself are on my "things to do before internship ends" and will likely happen sometime tomorrow

File: libquery/include/querying/ExpressionTree.h (hunk 19 above)
Comment (roshandathathri) [reply libquery/include/querying/ExpressionTree.h:31]: Neither did I.  But we treat all warnings as errors, so there shouldn't be a difference right?

Anyway, I would prefer classed types, it's cleaner to read.

**Comment by ddn0** [comment]
> and yes, performance wasn't really on our mind when we designed it this way. in particular, we did have some concerns about the way variables are being bound in the expression trees, but it looks like there's more than that

I don't mind landing this work but I do want to make sure we do a sanity check at some point to confirm that this is the architecture we want to continue with.

> some tests to test the filter op's operation itself are on my "things to do before internship ends" and will likely happen sometime tomorrow

Yeah. By tests, I mean filter op not the parser.

**Comment by l-hoang** [comment]
> Yeah. By tests, I mean filter op not the parser.

these current tests do test the backend "functionality" of the filter op (e.g., given an expression tree, evaluate it to get true or false) and not just the parsing, so i'm fairly confident that if filter op gets the correct expression tree then it's fine

the thing to test, then, is making sure filter op gets the right expression tree + filters rows correctly

File: libquery/src/ExpressionTree.cpp (hunk 18 above)
Comment (l-hoang) [reply libquery/src/ExpressionTree.cpp:73]: turns out you can't do this

templates being templates means it will create the code for float % float, and that doesn't work; the current approach will have to stay then

Comment (roshandathathri) [reply libquery/src/ExpressionTree.cpp:73]: Thanks for trying it out.

File: libquery/include/querying/ExpressionTree.h (hunk 19 above)
Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:31]: tried this, but it didn't work because I'm using the enum as the index name for the variants as well for readability purposes

e.g., std::get<kEAString>(some variant type)

Comment (l-hoang) [reply libquery/include/querying/ExpressionTree.h:31]: I guess I would have to do something like this:

https://stackoverflow.com/questions/8357240/how-to-automatically-convert-strongly-typed-enum-into-int

this would make readability quite bad, so i'm not in favor of this change

File: libquery/include/querying/ops/FilterOp.h (hunk 21 above)
Comment (l-hoang) [reply libquery/include/querying/ops/FilterOp.h]: i'm assuming that that was the fix, so i made the change; let me know if it was wrong and i can fix it

thanks for the catch

Comment (l-hoang) [reply libquery/include/querying/ops/FilterOp.h]: apparently that wasn't the way to go because clang complains

....i think i'll just move to manually moving the pointers

**Comment by l-hoang** [comment]
2020-08-28T01:36:06.1341578Z The following tests FAILED:
2020-08-28T01:36:06.1342261Z 	262 - run-bfs-push-dist-rmat15-async-cpu-oec-2-1-store_part (Failed)
2020-08-28T01:36:06.1342893Z 	263 - verify-bfs-push-dist-rmat15-async-cpu-oec-2-1-store_part (Not Run)

Seems like these spuriously fail sometimes; this isn't the first time I've noticed this.

**Comment by ddn0** [comment]
> Seems like these spuriously fail sometimes; this isn't the first time I've noticed this.

If this is during a sanitizer build, I think I know the cause, but if I'm right, it is already fixed upstream in libc++ so I didn't bother applying the workaround.

**Comment by l-hoang** [comment]
> > Seems like these spuriously fail sometimes; this isn't the first time I've noticed this.
> 
> If this is during a sanitizer build, I think I know the cause, but if I'm right, it is already fixed upstream in libc++ so I didn't bother applying the workaround.

I think it wasn't a sanitizer build IIRC.
